Bearded.UI/Events/EventRouter.cs

```diff
@@ -42,12 +36,60 @@ public static EventPropagationPath FindPropagationPath(IControlParent root, Cont
                     return EventPropagationPath.Empty;
                 }
                 path.Add(parentAsControl);
-                
+
                 parent = parentAsControl.Parent;
             }
 
             path.Reverse();
             return new EventPropagationPath(path.AsReadOnly());
         }
+
+        private static IEnumerable<Control> findPropagationPathUsingTest(
+            IControlParent root, PropagationTest propagationTest)
+        {
+            var path = new List<Control>();
+            findPropagationPathUsingTest(root, propagationTest, path);
+            return path;
+        }
+
+        private static bool findPropagationPathUsingTest(
+            IControlParent root, PropagationTest propagationTest, List<Control> path)
+        {
+            foreach (var child in root.Children.Reverse())
+            {
+                var outcome = propagationTest(child);
+                switch (outcome)
+                {
+                    case PropagationTestOutcome.Miss:
+                        break;
+                    case PropagationTestOutcome.PassThrough:
+                        if (child is IControlParent childAsPassThroughParent
+                            && childAsPassThroughParent.Children.Count > 0)
```

Comment:
Why this condition? It'll fail in the next recursive call anyways. Just to avoid one Add/RemoveAt pair?

Reply:
Removed it for readability reasons.